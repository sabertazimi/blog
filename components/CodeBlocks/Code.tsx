import { classNames } from '@components/utils';
import type { ReactNode } from 'react';
import styles from './Code.module.css';

interface Props {
  className?: string;
  children?: ReactNode;
}

const Code = ({ children, className, ...props }: Props): JSX.Element =>
  className?.includes('language-') ? (
    <code {...props} className={className}>
      {children}
    </code>
  ) : (
    <code {...props} className={classNames(className, styles.inline)}>
      {children}
    </code>
  );

export default Code;
