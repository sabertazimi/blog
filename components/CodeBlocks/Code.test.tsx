import { render } from '@testing-library/react';
import Code from './Code';

describe('Code', () => {
  const languages = [
    'html',
    'xml',
    'yml',
    'yaml',
    'css',
    'json',
    'md',
    'markdown',
    'js',
    'javascript',
    'ts',
    'typescript',
    'coffee',
    'coffeescript',
    'jsx',
    'tsx',
    'objc',
    'objectivec',
    'vue',
    'go',
    'java',
    'rust',
  ];

  test('should render inline code correctly (snapshot)', () => {
    const { container } = render(<Code>Inline Code</Code>);

    expect(container).toMatchSnapshot();
  });

  test('should render block code correctly (snapshot)', () => {
    const { container } = render(<Code className="language-type" />);

    expect(container).toMatchSnapshot();
  });

  test('should render live code correctly (snapshot)', () => {
    const { container } = render(
      <Code live={true} className="language-type">
        Live Code
      </Code>
    );

    expect(container).toMatchSnapshot();
  });

  test('should hidden line number and copy button according code block metadata (snapshot)', () => {
    const { container } = render(
      <Code noline={true} nocopy={true} className="language-type">
        const foo = bar();
      </Code>
    );

    expect(container).toMatchSnapshot();
  });

  test.each(languages)(
    'should render different language correctly (snapshot)',
    language => {
      const { container } = render(
        <Code className={`language-${language}`}>const foo = bar();</Code>
      );

      expect(container).toMatchSnapshot();
    }
  );
});
