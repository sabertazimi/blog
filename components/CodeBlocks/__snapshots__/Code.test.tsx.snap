// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code should render block code correctly (snapshot) 1`] = `
<div>
  <pre
    class="prism-code language-type block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        Block Code
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 1`] = `
<div>
  <pre
    class="prism-code language-html block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 2`] = `
<div>
  <pre
    class="prism-code language-xml block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 3`] = `
<div>
  <pre
    class="prism-code language-yml block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 4`] = `
<div>
  <pre
    class="prism-code language-yaml block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 5`] = `
<div>
  <pre
    class="prism-code language-css block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = 
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 6`] = `
<div>
  <pre
    class="prism-code language-json block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 7`] = `
<div>
  <pre
    class="prism-code language-md block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 8`] = `
<div>
  <pre
    class="prism-code language-markdown block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 9`] = `
<div>
  <pre
    class="prism-code language-js block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 10`] = `
<div>
  <pre
    class="prism-code language-javascript block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 11`] = `
<div>
  <pre
    class="prism-code language-ts block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 12`] = `
<div>
  <pre
    class="prism-code language-typescript block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 13`] = `
<div>
  <pre
    class="prism-code language-coffee block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 14`] = `
<div>
  <pre
    class="prism-code language-coffeescript block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 15`] = `
<div>
  <pre
    class="prism-code language-jsx block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 16`] = `
<div>
  <pre
    class="prism-code language-tsx block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 17`] = `
<div>
  <pre
    class="prism-code language-objc block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 18`] = `
<div>
  <pre
    class="prism-code language-objectivec block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 19`] = `
<div>
  <pre
    class="prism-code language-vue block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 20`] = `
<div>
  <pre
    class="prism-code language-go block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 21`] = `
<div>
  <pre
    class="prism-code language-java block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render different language correctly (snapshot) 22`] = `
<div>
  <pre
    class="prism-code language-rust block"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="line-number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`Code should render inline code correctly (snapshot) 1`] = `
<div>
  <code
    class="inline"
  >
    Inline Code
  </code>
</div>
`;
